### PR TITLE
Projects/ViTDet

### PR DESCRIPTION
**Pull request by zytx121**

### CODE REVIEW

File: projects/ViTDet/vitdet/convModule_norm.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    
+    def forward(self, x, activate=True, norm=True):
+        for layer in self.order:
+            if layer == 'conv':
+                if self.with_explicit_padding:
+                    x = self.padding_layer(x)
+                x = self.conv(x)
+            elif layer == 'norm' and norm and self.with_norm:
+                if 'LN' in self.normType:
+                    x = x.permute(0, 2, 3, 1)
```
Comment (ZwwWayne) [inline projects/ViTDet/vitdet/convModule_norm.py:21]: We can implement a new LN layer that does the permute and normalization, then choose to use the new LN layer through norm_cfg. In this way, we can reduce most of the code that needs to use ConvModule.

File: projects/ViTDet/vitdet/custom_layer_norm.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    """Custom Layer Normalization."""
+
+    def forward(self, x: Tensor) -> Tensor:
+        if x.dim() == 4:
+            x = x.permute(0, 2, 3, 1)
+            x = super().forward(x)
+            x = x.permute(0, 3, 1, 2).contiguous()
+        else:
+            x = super().forward(x)
+        return x
```
Comment (nijkah) [inline projects/ViTDet/vitdet/custom_layer_norm.py:17]: How about using this https://github.com/open-mmlab/mmclassification/blob/3d4f80d63c6ac929d487a512ee2cf0cd53c47ed8/mmcls/models/backbones/convnext.py#L19-L20 if this project supposes `MMClassificatoin` is already installed.

Comment (zytx121) [reply projects/ViTDet/vitdet/custom_layer_norm.py:17]: We also need to support x.dim() == 3 in this PR.

Comment (nijkah) [reply projects/ViTDet/vitdet/custom_layer_norm.py:17]: I understand. Thank you!

Comment (zytx121) [reply projects/ViTDet/vitdet/custom_layer_norm.py]: Good idea, we will take your advice.

File: projects/ViTDet/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+You should claim whether this is based on the pre-trained weights, which are converted from the official release; or it's a reproduced result obtained from retraining the model in this project. -->
+
+Here we provide the baseline version of ViTDet with ViT-base backbone.
+
+To find more variants, please visit the [official model zoo](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet#pretrained-models).
+
+|                                             Method                                              |  Backbone   | Pretrained Model |  Training set  |   Test set   | #epoch | box AP |         Download         |
+| :---------------------------------------------------------------------------------------------: | :---------: | :--------------: | :------------: | :----------: | :----: | :----: | :----------------------: |
+| [Faster R-CNN dummy](configs/mask-rcnn_vit-base_fpn_rpn-2conv_4conv1fc_ln-all_lsj-100e_coco.py) | DummyResNet |        -         | COCO2017 Train | COCO2017 Val |   12   | 0.8853 | [model](<>) \| [log](<>) |
```
Comment (ZwwWayne) [inline projects/ViTDet/README.md:52]: update the performance when benchmark is done.
We should also make sure it is consistent with D2.

**Comment by austinmw** [comment]
Hi, if you have time, I'm curious how I can try out this PR? I tried running the config, but got several registry scope errors. For example:

```
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:0,algo-1]<stdout>:02/08 11:34:31 - mmengine - WARNING - The "visualizer" registry in mmdet did not set import location. Fallback to call `mmdet.utils.register_all_modules` instead.
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:0,algo-1]<stdout>:02/08 11:34:31 - mmengine - WARNING - The "vis_backend" registry in mmdet did not set import location. Fallback to call `mmdet.utils.register_all_modules` instead.
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:0,algo-1]<stdout>:02/08 11:34:32 - mmengine - WARNING - The "model" registry in mmdet did not set import location. Fallback to call `mmdet.utils.register_all_modules` instead.
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:Traceback (most recent call last):
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmengine/registry/build_functions.py", line 121, in build_from_cfg
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    obj = obj_cls(**args)  # type: ignore
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/mmdetection/mmdet/models/necks/fpn.py", line 120, in __init__
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    l_conv = ConvModule(
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmcv/cnn/bricks/conv_module.py", line 148, in __init__
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    self.norm_name, norm = build_norm_layer(
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmcv/cnn/bricks/norm.py", line 107, in build_norm_layer
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    raise KeyError(f'Cannot find {norm_layer} in registry under scope '
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:KeyError: 'Cannot find None in registry under scope name mmdet'
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:During handling of the above exception, another exception occurred:
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:Traceback (most recent call last):
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmengine/registry/build_functions.py", line 121, in build_from_cfg
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    obj = obj_cls(**args)  # type: ignore
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/mmdetection/mmdet/models/detectors/mask_rcnn.py", line 22, in __init__
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    super().__init__(
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/mmdetection/mmdet/models/detectors/two_stage.py", line 37, in __init__
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    self.neck = MODELS.build(neck)
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmengine/registry/registry.py", line 537, in build
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    return self.build_func(cfg, *args, **kwargs, registry=self)
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmengine/registry/build_functions.py", line 240, in build_model_from_cfg
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    return build_from_cfg(cfg, registry, default_args)
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:  File "/opt/conda/lib/python3.9/site-packages/mmengine/registry/build_functions.py", line 135, in build_from_cfg
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:    raise type(e)(
bwg3ffjd50-algo-1-dqro5 | [1,mpirank:1,algo-1]<stderr>:KeyError: "class `SimFPN` in projects/ViTDet/vitdet/sim_fpn.py: 'Cannot find None in registry under scope name mmdet'"
```

**Comment by austinmw** [comment]
Hey, apologies since I know this PR is still a work in progress, but if I may ask, when modifying `img_size` (for example from 1024 to 448), what additional changes are necessary?

**Comment by ZwwWayne** [comment]
> Hey, apologies since I know this PR is still a work in progress, but if I may ask, when modifying `img_size` (for example from 1024 to 448), what additional changes are necessary?

You only need to modify the training/testing pipeline and the batch augments. But we have not tried before. Recently @zwhus has made some updates and the training accuracy is almost aligned. We will update this PR soon.

**Comment by ZwwWayne** [comment]
Closed and continued in #9812

**Pull request closed without merging**